### PR TITLE
feat(headers): support Opaque origin headers

### LINKED ISSUE #1065: Support Opaque origin headers, serializing it to `null`
Support Opaque origins:

> An internal value, with no serialization it can be recreated from (it is serialized as "null" per ASCII serialization of an origin), for which the only meaningful operation is testing for equality.

https://html.spec.whatwg.org/multipage/browsers.html#concept-origin

### PR DESCRIPTION
**Pull request by alexeyzab**
Add support for Opaque origin header, serializing it to `null`.
https://html.spec.whatwg.org/multipage/browsers.html#concept-origin

Closes #1065

- [x] The commit messages match the guidelines in https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md#git-commit-guidelines

### CODE REVIEW

File: src/header/common/origin.rs [hunk 1]
```diff
@@ -31,41 +31,70 @@ use header::parsing::from_one_raw_str;
...
     /// ```
     /// use hyper::header::Origin;
     /// let origin = Origin::new("https", "foo.com", Some(443));
-    /// assert_eq!(origin.scheme(), "https");
+    /// assert_eq!(origin.scheme(), Some("https"));
     /// ```
-    pub fn scheme(&self) -> &str {
-        &(self.scheme)
+    #[allow(unused_variables)]
+    // Variable `host` is considered unused, although there does not seem to be a way to omit it here.
```
Comment (seanmonstar) [inline src/header/common/origin.rs]: You can just elide all the fields you don't want, like so `OriginOrNull::Origin { ref scheme, .. }`

File: src/header/common/origin.rs [hunk 2]
```diff
@@ -31,41 +31,70 @@ use header::parsing::from_one_raw_str;
...
     /// use hyper::header::{Origin,Host};
     /// let origin = Origin::new("https", "foo.com", Some(443));
-    /// assert_eq!(origin.host(), &Host::new("foo.com", Some(443)));
+    /// assert_eq!(origin.host(), Some(&Host::new("foo.com", Some(443))));
     /// ```
-    pub fn host(&self) -> &Host {
-        &(self.host)
+    #[allow(unused_variables)]
+    // Variable `scheme` is considered unused, although there does not seem to be a way to omit it here.
+    pub fn host(&self) -> Option<&Host> {
```
Comment (seanmonstar) [inline src/header/common/origin.rs]: Aw, having to use `Option`s is unfortunate, that I had forgotten about. Oh well, there's no other way!

File: src/header/common/origin.rs (hunk 1 above)
Comment (alexeyzab) [inline src/header/common/origin.rs]: Oh, this is really neat, thanks for pointing it out! I am a bit new to Rust and didn't know about this until now.

Comment (jdm) [inline src/header/common/origin.rs]: Another trick is to add `_` to the names of otherwise-unused variables, which suppresses the warning.

**Comment by alexeyzab** [comment]
I went ahead and squashed the extra commit as well as force pushed it. 

Not sure if that's the right approach here, but I wanted to keep this change as a single commit. Let me know if there are other things I can add/improve here.

Thank you!

**Comment by seanmonstar** [comment]
Thanks! Excellent work!

**Pull request merged**